### ISSUE TITLE
Data Dependency Analysis for Aggregation

### ISSUE THREAD

**Issue by thomasrolinger** [thomasrolinger/chapel#35]
To know whether we can apply aggregation to a given statement in a loop, we need to consider data dependencies. To be conservative, we will say any access to the array of interest AFTER the aggregated statement is a data hazard. Anything before the statement is OK, since we assume iterations are order independent.

Unfortunately, this means some of our micro kernels cannot fully utilize aggregation since they have back-to-back statements on the same array that could be aggregated. But the above decision means only the last such statement would be aggregated.

Something we need to consider are the cases where the operation is in a `foreach` loop that is in the `forall`. We must check for invalid statements after the aggregated statement within the `foreach`, and then those after the `foreach`.

It comes down to being able to "know" if a given statement comes after the aggregated statement. I think we did something like this when creating the stripped down inspector.

We also need to do alias analysis to find statements mentioned above that make references to the arrays and then access them. Any aliases created after the aggregated statement do not need to be tracked down, as the simple statement `ref foo = A` would invalidate the optimization as an access to `A`. However, such a statement could be before the aggregated statement. In that case, we need to perform the same checks we do for `A` but for its alias (and any alias of that, etc.).

For functions, if we see that `A` is passed into a function call, we treat that as an access; there is no need to dive into the function and see what's happening. But if `A` is a global variable, we would still want to dive into functions to see if it is used. That should be easy enough to check by looking at all SymExprs involving `A`, detect if it is in a function and if that function is called after the aggregated statement within the loop. Or we could look for function calls after the aggregated statement, go into that function and look for uses of `A`. I believe the second approach will be easier in our situation.

Do we also need to think about accesses to `B`? In this case, we're dealing with issuing a read within the `A[B[i]]` access. When we call `agg.bufferOp(A[B[i]], val)`, we resolve the value at `B[i]`. Therefore, if we were to write to `B[i]` after this statement, it wouldn't change anything with the statement we are aggregating. And read-after-read is not a hazard. So I would say that we do not need to worry about what happens to `B` within the `forall` loop.

**Comment by thomasrolinger** [comment]
1. **Correctly identify RAW and WAW hazards in a simple `forall` loop.** Don't worry about diving into functions or handling aliases yet. We should be able to do this at normalization. One approach is to take the loop body and iterate through its statements via `for_alist(node, forall->loopBody())`. We find the statement we are aggregating and set a flag that will then invalidate the optimization if any subsequent statement involves `A`.
2. **Hande function calls.** If `A` is passed into the call, we invalidate the optimization as it indicates `A` is used in the call; that would be handled by (1) without really considering whether we have a function call or not. However, if `A` is NOT defined within the function that contains `forall` of interest, then it could be global and be accessed in the called function without being passed in as an argument. In that case, when we do the checks in (1) and the statement we come across is a function call, then we look for any SymExprs that involve `A` within that function. We should do this in a new function that takes in a `BlockStmt` to look at. That way, we can recursively do these checks if we find other calls.
3. **Handle aliases.** For now, this should be handled by (1) since something like `ref foo = A` should be caught and marked as invalid. We could go further and ensure that `foo` is used before invalidating, since if it isn't used then there is no problem.

**Comment by thomasrolinger** [comment]
This is now implemented and working: https://github.com/thomasrolinger/chapel/commit/de170892f37e6d76f5f0713fc14ab7fbc55c22c5

It went more or less as described above. We first find all aliases to the array `A` and store them in a set. We then start at the top of the `forall` body and use an `for_alist` loop to arrive at the aggregation call/statement. Once we reach that, we then process each subsequent node in the `for_alist` loop and check whether any of the alias symbols are used. The only special case here is detecting function calls that do not take in the symbols as arguments. We just dive into the function body and look for uses there (and do so recursively).

All of the real-world applications are working as intended and I ensured that we correctly catch data dependencies that were added specifically to some small test cases.
